dev-client/src/model/sync/sync.ts

```diff
@@ -0,0 +1,144 @@
+/*
+ * Copyright © 2024 Technology Matters
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published
+ * by the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program. If not, see https://www.gnu.org/licenses/.
+ */
+
+export type ChangeRecords<T> = Record<string, ChangeRecord<T>>;
+
+export type ChangeTimestamp = number;
+
+export type ChangeRevisionId = number;
+
+export type ChangeRecord<T> = {
+  /**
+   * Unique ID for the current state since the last sync, monotonically increasing for each change.
+   * Can be used to determine whether sync responses reflect stale data.
+   * Cleared for a successful sync.
+   * Default to zero for uninitialized records.
+   */
+  revisionId?: ChangeRevisionId;
+  lastModifiedAt?: ChangeTimestamp;
+
+  lastSyncedRevisionId?: ChangeRevisionId;
+  lastSyncedData?: T;
+  lastSyncedAt?: ChangeTimestamp;
+};
+
+export const INITIAL_REVISION_ID = 0;
+
+export const getRevisionId = <T>(
+  record?: ChangeRecord<T>,
+): ChangeRevisionId => {
+  return record?.revisionId ?? INITIAL_REVISION_ID;
+};
+
+export const nextRevisionId = (
+  revisionId: ChangeRevisionId,
+): ChangeRevisionId => {
+  return revisionId + 1;
+};
+
+export const getChanges = <T>(
+  records: ChangeRecords<T>,
+  ids: string[],
+): ChangeRecords<T> => {
+  return Object.fromEntries(ids.map(id => [id, getChange(records, id)]));
+};
+
+export const getChange = <T>(
+  records: ChangeRecords<T>,
+  id: string,
+): ChangeRecord<T> => {
+  return records[id] ?? {};
+};
+
+export const markAllChanged = <T>(
+  records: ChangeRecords<T>,
+  ids: string[],
+  at: ChangeTimestamp,
+) => {
+  for (const id of ids) {
+    markChanged(records, id, at);
+  }
+};
+
+export const markChanged = <T>(
+  records: ChangeRecords<T>,
+  id: string,
+  at: ChangeTimestamp,
+) => {
+  const prevRecord = getChange(records, id);
+  const prevRevisionId = getRevisionId(prevRecord);
+  const revisionId = nextRevisionId(prevRevisionId);
+  records[id] = {
+    revisionId: revisionId,
+    lastModifiedAt: at,
+
+    lastSyncedRevisionId: prevRecord?.lastSyncedRevisionId,
+    lastSyncedData: prevRecord?.lastSyncedData,
+    lastSyncedAt: prevRecord?.lastSyncedAt,
+  };
+};
```

Comment:
question: am i remembering right that this version is meant to handle cases where the client makes further changes while a push or pull is in flight? i'm having trouble remembering the logic of how/in what order these methods are going to be called to handle that scenario (or am maybe misremembering whether this version is supposed to handle that scenario) and am wondering if you could write it up in a comment somewhere in this file